### PR TITLE
[mlir] Add inferContractionDims util for indexing map inputs

### PR DESCRIPTION
**Pull request by Max191**
This PR adds a util function to infer contraction dimensions given only the indexing maps of a linalg operation.

### CODE REVIEW

File: mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp [hunk 1]
```diff
@@ -201,11 +193,48 @@ findPermutationsIndexingOperand(LinalgOp linalgOp, OpOperand *opOperand,
...
+findPermutationsIndexingOperand(LinalgOp linalgOp, OpOperand *opOperand,
+                                utils::IteratorType iter) {
+  assert(linalgOp == opOperand->getOwner() && "expected linalgOp owner");
+  return findPermutationsIndexingOperandImpl(
+      linalgOp.getMatchingIndexingMap(opOperand),
+      linalgOp.getIteratorTypesArray(), iter);
+}
+
+static llvm::SmallDenseSet<int64_t>
+findPermutationsIndexingOperand(AffineMap indexingMap,
```
Comment (qedawkins) [inline mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp:213]: Do we need this static function? I think you can just call the impl directly where appropriate.

File: mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp [hunk 2]
```diff
@@ -201,11 +193,48 @@ findPermutationsIndexingOperand(LinalgOp linalgOp, OpOperand *opOperand,
...
 namespace {
 auto par = utils::IteratorType::parallel;
 auto red = utils::IteratorType::reduction;
 } // namespace
 
+/// Infer the iterator types from the init affine map. This looks at which dims
+/// are present in the map results, and returns an iterator types array with
+/// parallel types for dims that are present, and reduction types for dims that
+/// are not present.
+static ArrayRef<utils::IteratorType> inferIteratorsFromOutMap(AffineMap map) {
```
Comment (qedawkins) [inline mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp:228]: nit: Given that inferContractionDims is an API function, it might be nice to make this `FailureOr` and fail if not a projected permutation.

**Review by qedawkins — CHANGES_REQUESTED** [review]

File: mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp (hunk 1 above)
Comment (Max191) [reply mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp:213]: I personally prefer not to call the impl function and just use overloads, but I can change it if you prefer me to directly call the impl given that it is just a static helper.

Comment (qedawkins) [reply mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp:215]: Maybe just `findPermutationsIndexingOperandImpl` -> `findPermutationsIndexingOperand` then and drop this?

File: mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp [hunk 3]
```diff
@@ -201,11 +202,47 @@ findPermutationsIndexingOperand(LinalgOp linalgOp, OpOperand *opOperand,
...
 namespace {
 auto par = utils::IteratorType::parallel;
 auto red = utils::IteratorType::reduction;
 } // namespace
 
+/// Infer the iterator types from the init affine map. This looks at which dims
+/// are present in the map results, and returns an iterator types array with
+/// parallel types for dims that are present, and reduction types for dims that
+/// are not present.
+static FailureOr<ArrayRef<utils::IteratorType>>
```
Comment (qedawkins) [inline mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp:231]: nit: I would just return the SmallVector directly

File: mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp [hunk 4]
```diff
@@ -201,11 +202,47 @@ findPermutationsIndexingOperand(LinalgOp linalgOp, OpOperand *opOperand,
...
+  if (!map.isProjectedPermutation())
+    return failure();
+  SmallVector<utils::IteratorType> iterators(map.getNumDims(), red);
+  for (auto expr : map.getResults()) {
+    if (auto dim = dyn_cast<AffineDimExpr>(expr)) {
+      iterators[dim.getPosition()] = par;
+    }
+  }
+  if (iterators.size() != map.getNumDims())
+    return failure();
```
Comment (qedawkins) [inline mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp:242]: nit: This check does not seem possible given you aren't changing the size of the vector.

**Review by qedawkins — APPROVED** [review]
Ok this looks fine to me, but given there is no test here and it's just changing API surface, I'm not an authority on what to/how to make API changes like this so give others a chance to review before landing.

File: mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp (hunk 3 above)
Comment (chelini) [inline mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp:231]: Is this method correct? ArrayRef does not own the underlying data and you return the small vector allocated in the function.

File: mlir/include/mlir/Dialect/Linalg/IR/LinalgInterfaces.h [hunk 5]
```diff
@@ -62,6 +62,8 @@ struct ContractionDimensions {
 /// `k`, indices are returned in sorted order.
 /// Returns a failure if any of `m`, `n` or `k` is empty.
 FailureOr<ContractionDimensions> inferContractionDims(LinalgOp linalgOp);
+FailureOr<ContractionDimensions>
```
Comment (chelini) [inline mlir/include/mlir/Dialect/Linalg/IR/LinalgInterfaces.h:65]: Out of curiosity, why is this something useful? What use case do you have in mind?

Comment (Max191) [reply mlir/include/mlir/Dialect/Linalg/IR/LinalgInterfaces.h:65]: I need it here: https://github.com/openxla/iree/pull/15984
TLDR, We have an attribute on types tracking the indexing maps, so we don't have access to the actual op.

But this can also be useful more generally in cases where you only have access to indexing maps. For example, doing indexing map analysis before creating a new op. It seems like a potentially useful util to have to me.

File: mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp (hunk 3 above)
Comment (Max191) [reply mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp:231]: It seems to be working for me, but probably better to return SmallVector. I'll update

File: mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp [hunk 6]
```diff
@@ -15,9 +15,11 @@
 #include "mlir/Dialect/Linalg/IR/Linalg.h"
 #include "mlir/Dialect/MemRef/IR/MemRef.h"
 #include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Utils/StructuredOpsUtils.h"
```
Comment (chelini) [inline mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp:18]: Are these includes needed?

File: mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp [hunk 7]
```diff
@@ -201,11 +203,37 @@ findPermutationsIndexingOperand(LinalgOp linalgOp, OpOperand *opOperand,
   return res;
 }
 
+static llvm::SmallDenseSet<int64_t>
```
Comment (chelini) [inline mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp:206]: I would drop this method. It is used only for `inferConvolutionDimsImpl`.

File: mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp [hunk 8]
```diff
@@ -201,11 +203,37 @@ findPermutationsIndexingOperand(LinalgOp linalgOp, OpOperand *opOperand,
...
+/// Infer the iterator types from the init affine map. This looks at which dims
+/// are present in the map results, and returns an iterator types array with
+/// parallel types for dims that are present, and reduction types for dims that
+/// are not present.
+static FailureOr<SmallVector<utils::IteratorType>>
+inferIteratorsFromOutMap(AffineMap map) {
+  if (!map.isProjectedPermutation())
+    return failure();
+  SmallVector<utils::IteratorType> iterators(map.getNumDims(), red);
+  for (auto expr : map.getResults()) {
```
Comment (chelini) [inline mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp:229]: nit: drop braces for simple single stmt body if and for.

File: mlir/include/mlir/Dialect/Linalg/IR/LinalgInterfaces.h (hunk 5 above)
Comment (chelini) [reply mlir/include/mlir/Dialect/Linalg/IR/LinalgInterfaces.h:65]: Ok, thanks. I don't have a strong opinion here, but I would expect something more generic if we go down to affine maps (i.e., it is unclear why there is a restriction on the number of affine maps).

Comment (Max191) [reply mlir/include/mlir/Dialect/Linalg/IR/LinalgInterfaces.h:65]: Given the definition of the ContractionOpInterface, I think it makes sense to keep it restricted to 3 maps (lhs, rhs, out) https://github.com/llvm/llvm-project/blob/74a09bd1ec6066d56880df1ae1a2c0258442cee9/mlir/include/mlir/Dialect/Linalg/IR/LinalgInterfaces.td#L19-L35
Is there a more general form of a tensor contraction that you are thinking of? Perhaps something like that could live elsewhere to avoid confusion with the `ContractionOpInterface` defined here.

**Review by hanhanW — APPROVED** [review]
LGTM, thank you

File: mlir/include/mlir/Dialect/Linalg/IR/LinalgInterfaces.h (hunk 5 above)
Comment (chelini) [reply mlir/include/mlir/Dialect/Linalg/IR/LinalgInterfaces.h:65]: Sorry for the latency, I lost this PR. Yes, you are right `ContractionOpInterface` already assumes two inputs and one output. It makes sense to restrict to 3 maps only then.

**Review by chelini — APPROVED** [review]
